crates/wasm-encoder/src/lib.rs

```diff
@@ -209,31 +237,79 @@ impl Default for Module {
     }
 }
 
-/// Known section IDs.
+/// Represents a WebAssembly component that is being encoded.
 ///
-/// Useful for implementing the `Section` trait, or for setting
-/// `RawSection::id`.
+/// Unlike core WebAssembly modules, the sections of a component
+/// may appear in any order and may be repeated.
+#[derive(Clone, Debug)]
+pub struct Component {
```

Comment:
My personal preference would be to get to the point where we can do away with this type. It looks like the component/module distinction is purely used for assertions right now and doesn't actually otherwise affect encoding (beyond the version byte). In that sense I commented above about how I personally think that the assertions aren't necessary, and if we go that route then I think there's room for:

```rust
impl Module {
    fn new() -> Module; // creates a core wasm module
    fn component() -> Module; // creates a component module
    // ...
}
```

I think that would also allow the removal of `ComponentSection`? There might need to be a new layer of abstraction for encoding the section id where some section ids all encode as the same value if the value clashes between core wasm and not, but I don't think that poses an API issue for this crate.

Basically I feel that this is a good opportunity to have core wasm and the component model be closely aligned in terms of API hierarchy and encourage maximal reuse between the two.

Reply:
If we can align the section ids, I will remove `ComponentSection` and `Component` entirely.